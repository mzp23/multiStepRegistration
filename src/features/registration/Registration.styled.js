import styled from "styled-components";

export const Wrapper = styled.div`
  padding-top: 15%;
  height: 100vh;
`;
export const Title = styled.h1`
  color: #a1a1a1;
  text-align: center;
  margin-bottom: 10px;
`;

export const RegistrationWrapper = styled.div`
display: flex;
flex-direction: column;
max-width: 400px;
margin: 0 auto;
`;

export const InputContainer = styled.div`
  display: flex;
  align-items: baseline;
`;

export const Label = styled.label`
  width: 30%;
`;