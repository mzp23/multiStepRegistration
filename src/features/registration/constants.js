export const FIRST_STEP = 1;
export const SECOND_STEP = 2;
export const LAST_STEP = 3;
