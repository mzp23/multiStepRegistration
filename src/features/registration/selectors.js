export const selectFirstName = ({ registration }) => registration.firstName;
export const selectSecondName = ({ registration }) => registration.secondName;
export const selectAge = ({ registration }) => registration.age;
export const selectEmail = ({ registration }) => registration.email;
export const selectDob = ({ registration }) => registration.dob;
export const selectConsentToProcessingData = ({ registration }) =>
  registration.consentToProcessingData;
export const selectSex = ({ registration }) => registration.sex;
