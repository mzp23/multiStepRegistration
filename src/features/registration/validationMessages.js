export const invalidFirstNameMessage = 'First Name should be greater than 2 and less than 16';
export const invalidSecondNameMessage = 'Second Name should be greater than 2 and less than 16';
export const invalidEmailMessage = 'Invalid email address';
export const invalidPasswordMessage = 'Password should be greater than 5';
export const invalidConfirmedPasswordMessage = 'Passwords do not match';
export const invalidZipCodeMessage = 'Enter valid zip code'