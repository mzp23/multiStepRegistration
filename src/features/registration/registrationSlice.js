import { createSlice } from '@reduxjs/toolkit';
import {
  isValidEmail,
  isValidFirstName,
  isValidPassword,
  isValidSecondName,
  isValidConfirmedPassword,
  isValidZipCode,
  isValidDob,
  isValidSex,
} from '../../helpers/validation';
import { FIRST_STEP, LAST_STEP } from './constants';

const initialState = {
  firstName: {
    value: '',
    isValid: false,
  },
  secondName: {
    value: '',
    isValid: false,
  },
  password: {
    value: '',
    isValid: false,
  },
  confirmedPassword: {
    value: '',
    isValid: false,
  },
  zipCode: {
    value: '',
    isValid: false,
  },
  email: {
    value: '',
    isValid: false,
  },
  sex: {
    value: '',
    isValid: false,
  },
  dob: {
    value: null,
    isValid: false,
  },
  consentToProcessingData: false,
  registrationStep: 1,
};

const validationMap = {
  firstName: isValidFirstName,
  secondName: isValidSecondName,
  email: isValidEmail,
  password: isValidPassword,
  confirmedPassword: isValidConfirmedPassword,
  zipCode: isValidZipCode,
  dob: isValidDob,
  sex: isValidSex,
};

export const registrationSlice = createSlice({
  name: 'registration',
  initialState,
  reducers: {
    changeFieldValue: (state, { payload }) => {
      state[payload.field].value = payload.value;

      if (payload.field === 'confirmedPassword') {
        state[payload.field].isValid = validationMap[payload.field](
          payload.value,
          state.password.value
        );
      } else {
        state[payload.field].isValid = validationMap[payload.field](
          payload.value
        );
      }
    },
    changeConsentToProcessingData: (state, { payload }) => {
      state.consentToProcessingData = payload;
    },
    reset: state => initialState,
    increaseStep: (state) => {
      if (state.registrationStep >= LAST_STEP) {
        state.registrationStep = LAST_STEP;
      } else {
        state.registrationStep = state.registrationStep + 1;
      }
    },
    decreaseStep: (state) => {
      if (state.registrationStep <= FIRST_STEP) {
        state.registrationStep = FIRST_STEP;
      } else {
        state.registrationStep = state.registrationStep - 1;
      }
    },
  },
});

export const {
  increaseStep,
  decreaseStep,
  changeFieldValue,
  changeConsentToProcessingData,
  reset,
} = registrationSlice.actions;

export default registrationSlice.reducer;
