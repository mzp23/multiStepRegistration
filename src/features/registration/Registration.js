import React from 'react';
import { useSelector } from 'react-redux';
import styled from 'styled-components';

import Buttons from '../../components/registration/buttons/Buttons';
import { LAST_STEP } from './constants';
import FirstStep from './firstStep/FirstStep';
import SecondStep from './secondStep/SecondStep';
import { selectStep } from './selectors';

const Wrapper = styled.div`
  padding-top: 15%;
  height: 100vh;
`;
const Title = styled.h1`
  color: #a1a1a1;
  text-align: center;
  margin-bottom: 10px;
`;
const Registration = () => {
  const step = useSelector(selectStep);
  const stepTitle = step === LAST_STEP ? 'Conformation' : `Step ${step}`;

  return (
    <Wrapper>
      <Title>{stepTitle}</Title>
      <FirstStep />
      <SecondStep />
      <Buttons />
    </Wrapper>
  );
};

export default Registration;
