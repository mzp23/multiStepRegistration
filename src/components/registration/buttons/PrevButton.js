import React, { useCallback } from 'react';
import { useDispatch } from 'react-redux';
import { Button } from 'rsuite';
import { decreaseStep } from '../../../features/registration/registrationSlice';

const PrevButton = () => {
  const dispatch = useDispatch();
  const handlePrevStep = useCallback(() => {
    dispatch(decreaseStep());
  }, [dispatch]);

  return (
    <Button type="button" onClick={handlePrevStep}>
      prev
    </Button>
  );
};

export default PrevButton;
