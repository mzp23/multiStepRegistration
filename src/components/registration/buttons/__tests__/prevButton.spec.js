import React from 'react';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';
import PrevButton from '../PrevButton';
import * as ReactRedux from 'react-redux';
import { decreaseStep } from '../../../../features/registration/registrationSlice';

const setUp = () => shallow(<PrevButton />);
const initialState = {
  registrationStep: 2,
};
const mockStore = configureStore();

describe('<PrevButton />', () => {
  const mockDispatch = jest.fn();
  let component;
  const store = mockStore(initialState);
  beforeAll(() => {
    ReactRedux.useDispatch = jest.fn().mockImplementation(() => mockDispatch);
  });

  beforeEach(() => {
    ReactRedux.useDispatch.mockClear();
    component = setUp();
  });
  it('should render PrevButton correct', () => {
    expect(component).toMatchSnapshot();
  });

  it('should call onClick callback', () => {
    component.simulate('click');
    expect(mockDispatch).toHaveBeenCalled();
    expect(mockDispatch).toHaveBeenCalledWith(decreaseStep());
  });
});
