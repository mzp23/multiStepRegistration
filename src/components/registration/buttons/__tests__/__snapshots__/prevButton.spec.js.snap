// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PrevButton /> should render PrevButton correct 1`] = `
<withStyleProps(defaultProps(Button))
  appearance="default"
  classPrefix="rs-btn"
  componentClass="button"
  onClick={[Function]}
  ripple={true}
  type="button"
>
  prev
</withStyleProps(defaultProps(Button))>
`;
